HopsanGUI/LogVariable.cpp

```diff
@@ -805,6 +805,130 @@ SharedVectorVariableT VectorVariable::toFrequencySpectrum(const SharedVectorVari
+                continue;
+            }
+
+            //Compute first variance
+            double xmax = data[i-nw];
+            double xmin = data[i-nw];
+            for(int j=i-nw+1; j<i; ++j) {
+                if(data[j] > xmax) {
+                    xmax = data[j];
+                }
+                else if(data[j] < xmin) {
+                    xmin = data[j];
+                }
+            }
+
+            if(xmax-xmin < tol) {
+                pRetVar->append(time[i], 1);
+            }
+            else {
+                pRetVar->append(time[i], 0);
+            }
+        }
+    }
+    else if(method == VarianceRatioTest) {
+
+        //Create randomized data vector
+        QVector<double> x;
+        x.resize(data.size());
+        for(int i=0; i<data.size(); ++i) {
+            x[i] = normalDistribution(data[i], stdev);
+        }
+
+        int nw = 0;
+        for(int i=0; i<time.size(); ++i) {
+            if(time[i] < win) {
+                pRetVar->append(time[i], 0);
+                nw = i+1;  //Update number of samples for window
+                continue;
+            }
+
+
+            //Compute average of window
+            double mean = 0;
+            for(int j=i-nw; j<i; ++j) {
+                mean += x[j];
+            }
+            mean /= nw;
+
+            //Compute first variance
+            double s1 = 0;
+            for(int j=i-nw; j<i; ++j) {
+                s1 += (x[j]-mean)*(x[j]-mean);
+            }
+            s1 /= (nw/*-1.0*/);
+
+            //Compute second variance
+            double s2 = 0;
+            for(int j=i-nw; j<i-1; ++j) {
+                s2 += (x[j+1]-x[j])*(x[j+1]-x[j]);
+            }
+            s2 /= (nw-1.0);
+
+            //Avoid division by zero
+            s1 = fmax(std::numeric_limits<double>::min(), fabs(s1));
+            s2 = fmax(std::numeric_limits<double>::min(), fabs(s2));
+
+            if(s1/s2 < tol) {
+                pRetVar->append(time[i], 1);
+            }
+            else {
+                pRetVar->append(time[i], 0);
+            }
+        }
+    }
+    else if(method == MovingAverageVarianceRatioTest) {
+        double xf = 0;
+        double df = 0;
+        double xold = data[0];
+        pRetVar->append(time[0], 0);
+        for(int i=1; i<time.size(); ++i) {
+            double x = normalDistribution(data[i], stdev);
+
+            double vf = l2*(x-xf)*(x-xf);
+
+            xf = l1*x+(1.0-l1)*xf;
+
+            double s1 = (2.0-l1)/2.0*vf;
+
+            df = l3*(x-xold)*(x-xold) + (1-l3)*df;
+            double s2 = df/2.0;
+
+            //Avoid division by zero
+            s1 = fmax(std::numeric_limits<double>::min(), fabs(s1));
+            s2 = fmax(std::numeric_limits<double>::min(), fabs(s2));
+
+            if(s1/s2 < tol) {
+                pRetVar->append(time[i], 1);
+            }
+            else {
+                pRetVar->append(time[i], 0);
+            }
+            xold = x;
+        }
+    }
```

Comment:
I know it would be time consuming, but for all of these mathematical algorithms it would be nice with some, comments explaining the algorithms, possibly with embedded equations (maybe latex). I dont know the math behind these algorithms so I can not properly review them. And in the future if someone looks at this code some explanation would be nice so that the intended behavior would be clear. If you can add something like that it would be good. (but not required). I guess an alternative would be to look up the methods on Wikipedia or similar.

Reply:
The problem is that the even if the theory is in the markdown file for the component, the user will not know where to look for it when trying to understand the function in HCOM. But that problem can't be easily solved I think.